### DISCUSSION TITLE
Is anyone else's Confidence levels way off vs BirdNET-Pi?

### DISCUSSION THREAD

**Discussion by kchristensen** [mmcc-xx/BirdCAGE#40] [General]
I've got this up and running to do a comparison to BirdNET-Pi (no contest really, this is far less janky) but I've found that using the same RTSP stream, that the confidence level on something like an American Robin which BirdNET-Pi can identify with 90%+ confidence is like 26% Confidence in BirdCAGE. I have my location set appropriately, etc. If I don't bump my confidence down from .7 to something like .2, it won't actually identify anything at all, which seems off.

**Comment by mmcc-xx** [comment]
>  this is far less janky

Thanks - that's what I was going for :)

> the confidence level on something like an American Robin which BirdNET-Pi can identify with 90%+ confidence is like 26% Confidence in BirdCAGE

That's odd. I haven't done a direct comparison, but I do get quite a few detections that exceed a 70% confidence.
 
* Are you using the same model in both installations? Last I checked, BirdNET-Pi was using an older model by default. Right now on in the GitHub repo they have the new V2.4 model, which you can use with this app by making a change in the docker-compose file.
* I'm not doing any pre-processing of the wav file before handing it to the analyzer. Do you know if BirdNET-Pi is doing any manipulation of the audio? Noise reduction or amplification or the like?
* Do you have your lat/long set correctly, with a negative number for longitude if you are in the western hemisphere and a negative number for latitude if you are south of the equator?

**Comment by kchristensen** [comment]
* I just switched yesterday on the birdnet installation to the new model, but even with the old model birds like the Robin are generally the confidence is super high like 90-95%. 
* I don’t think they are doing anything other than [taking a stream and dumping it to a file](https://github.com/mcguirepr89/BirdNET-Pi/blob/main/scripts/birdnet_recording.sh), the analysis script doesn’t seem to do much either as far as I can tell.
* Yup, verified my lat/lon is fine, I copied it over from BirdNET and triple checked it in google maps just to be sure.

An example entry from the logs:

```
{'12.0;15.0': [('Turdus migratorius_American Robin', 0.2615), ('Pheucticus ludovicianus_Rose-breasted Grosbeak', 0.0334)], '6.0;9.0': [('Piranga olivacea_Scarlet Tanager', 0.0238), ('Empidonax traillii_Willow Flycatcher', 0.0139), ('Icterus galbula_Baltimore Oriole', 0.0096)]}
```

I’m not exactly certain what the syntax is there, but I assume the `0.2615` is the confidence level, and it was like solid robins outside at the time of day that ran.

**Comment by kchristensen** [comment]
Here’s a snippet from BirdNET-Pi of some robin findings:

<img width="1194" alt="image" src="https://github.com/mmcc-xx/BirdCAGE/assets/838358/a502327c-9153-48bb-9004-9766a1240bcd">

**Comment by mmcc-xx** [comment]
Same "sh thresh" value for your location? I think in birdnet-pi it is set where you set your lat/long value.

There's a "sensitivity" parameter that is gets passed into the model - I don't know what it does. I have mine set to 1.2, but I don't remember why I set it to that :) Any idea if birdnet-pi uses that?

**Comment by kchristensen** [comment]
I've gone back and forth and as best I can tell, my settings between the two are now the same but the confidence coming back from the birdcage-server container is way lower. It's weird, if I lower my confidence threshold to like .2, it's not misidentifying birds I would normally have around.

**Reply by mmcc-xx** [reply]
One thing I noticed about the ffmpeg command used in BNPi is that it converts the audio coming in to 48KHz sample rate (if I’m remembering correctly)… I wonder if that has anything to do with it…

**Comment by kchristensen** [comment]
I tried 2 things yesterday, not sure if either really made a difference but I am seeing some findings now > .5 or so:

* I was using the low quality rtsp feed off my camera, so I switched to the high one. I have to imagine that doesn’t impact the audio at all since it’s probably the same feed regardless, but whatever.
* Related to https://github.com/mmcc-xx/BirdCAGE/discussions/41 I was annoyed that the timestamps displayed were in UTC since that’s what my server is on, but I am in EDT so I copied EDT versions of `/etc/localtime` and `/etc/timezone` into my birdcage data directory and mounted them into the containers.

I know location factors into the analysis, but I don’t think time of day does, but with my confidence at .6 I am at least getting some findings now. I am going to take a look later today and compare like, the number of Robins BirdNET-Pi found vs BirdCAGE and see how close we are.

**Reply by mmcc-xx** [reply]
That's odd... perhaps the sample rate somehow does matter?

The time of day doesn't get passed into the birdnet container, and as far as I know it doesn't look at the time of day. But maybe?

**Comment by mmcc-xx** [comment]
I'm running two instances on the same rtsp stream, but on one I'm forcing the recording to be saved as 16bit, 48KHz mono. So far I'm not seeing any significant difference in the detections but the spectrogram for the pretend high res one looks better. I'll probably add it in just to make the pictures nicer even if it doesn't have a difference on the detection quality.

**Comment by kchristensen** [comment]
I must have spoken too soon, because apparently since about 8 this morning I haven't had any detections with a confidence level > .5

![image](https://github.com/mmcc-xx/BirdCAGE/assets/838358/3b96cda1-0b52-47af-b9a2-c89043777d2f)

I'm going to have to dig in more but for the life of me I can't find anything different between the two settings wise.

**Comment by mmcc-xx** [comment]
I assume it was producing log files all day - just with results below your threshold?

If you open the stream in VLC and look at Tool -> Codec Information, what do you see?

I'm going to push images with the up-res-ing of the audio. Give that a try and see if you notice any difference.

**Comment by kchristensen** [comment]
Yeah the server is constantly finding birds, just with much lower thresholds. Oddly VLC won’t open any of my camera streams as they are coming out of UniFi Protect as RTSPS and VLC apparently hates that.

I just pulled the uprezzing code into my branch we’ll see what it does. I’m going to try to figure out what the bitrate on this stream is somehow.

**Comment by kchristensen** [comment]
Finally figured out how to get the stream playing in VLC, it seems as though it should be plenty high-quality enough:

![image](https://github.com/mmcc-xx/BirdCAGE/assets/838358/84d9f05a-4d33-4e15-b976-e919641f0a9d)

**Reply by mmcc-xx** [reply]
Yeah - who knows how they are using all of those bits but there are plenty of them. Maybe it needed to be down-sampled? The new images should up or down sample as needed to get it to 16bit/48KHz. Have you seen any difference?

**Comment by kchristensen** [comment]
With my confidence level at .5, I am definitely getting hits now, although I think there's probably some false positives (not seen in this screenshot) with it set that low. I'm going to slowly creep back towards .7 and see what happens. I still have zero reason to believe that it should be any different than BirdNET-Pi but we'll see.

![image](https://github.com/mmcc-xx/BirdCAGE/assets/838358/c5103f55-1896-4081-8e2e-c88067f1dd2f)

Related: Do you have plans to push up the server container bits sometime soon? When I was poking around this I was going to hack on that a bit but realized it's not actually in the repo yet.

**Reply by mmcc-xx** [reply]
The birdnetserver is the BirdNET-Analyzer repo with a couple of files changed. I think the changes are up to date here:

https://github.com/mmcc-xx/BirdCAGE/tree/main/BirdNET-Analyzer

I made some changes to the Dockerfile that haven't gotten uploaded.

I should just upload the whole damn thing I suppose. However, I'm leaving on a trip tomorrow and will be out of pocket for a couple weeks.

**Reply by kchristensen** [reply]
No worries, have a great vacation!
